### PR TITLE
Migrate telemetry pings to Glean (Stage 2)

### LINKED ISSUE #2230: Migrate Telemetry pings to Glean (Stage 2)
**Is your feature request related to a problem? Please describe.**
Due to the current telemetry service is gonna deprecated by 2020 Q1, so we need to migrate it to the new system, Glean, as soon as we can.

Per #1854, we have enabled Glean and supported its baseline pings. So, we gonna start to migrate other pings to Glean in this stage 2.

**Describe the solution you'd like**
In this stage, I would expect we can migrate all of our pings except for the telemetry that are related to multiple windows because we haven't been in the same page yet.

### PR DESCRIPTION
**Pull request by daoshengmu**
Fixes #2230 . Migrating the telemetry for

- [x] Counting how many domains and URLs are visited in a session.

- [x] How many times do users use url insert, search query, and voice query in the URL bar?

- [x] How many searches are queried in a specific search engine?

### CODE REVIEW

**Comment by daoshengmu** [comment]
Blocked by [Bug 1595914](https://bugzilla.mozilla.org/show_bug.cgi?id=1595914) to access `GleanTimerId`

File: app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java [hunk 1]
```diff
@@ -53,6 +69,100 @@ public static void stop() {
...
+
+    }
+
+    public static void sessionStop() {
+        domainMap.clear();
+
+        // TODO: Because we measure it by sessions,
+        // How to reset
+        // Url.INSTANCE.getDomains() and Url.INSTANCE.getVisits() counter to zero?
+    }
```
Comment (daoshengmu) [inline app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java]: @Dexterp37 . We have some metrics are measured by sessions, so we have to reset the `Counter` to zero after a session. Is there a way to do this?

Comment (Dexterp37) [reply app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java]: What is a `session`? There is no such thing as a `session` concept in the Glean world.

Metrics are automatically cleared depending on their [lifetime](https://mozilla.github.io/glean/book/user/metric-parameters.html#optional-metric-parameters).

The `metrics` ping, for example, is sent once per day at 4am (or other times, if it's overdue). If a metric has `lifetime: ping` (default), then its value is cleared automatically after the ping is assembled.

This means that you don't really need to care about this, unless you really want to measure things in a different cadence.

For the specific `domains` metric, since you want to measure the number of unique `domains` visited, this could be trickier. @chutten, thoughts?

File: app/metrics.yaml [hunk 2]
```diff
@@ -23,3 +23,152 @@ distribution:
...
       - dmu@mozilla.com
     expires: "2020-05-01"
+pages:
+  page_load:
+    type: timing_distribution
+    lifetime: application
+    description: >
+      Counting how long each page takes to load.
+    send_in_pings:
+      - metrics
```
Comment (Dexterp37) [inline app/metrics.yaml]: This is redundant, these metrics will go by default into the `metrics ping`. You can remove both lines.

File: app/metrics.yaml [hunk 3]
```diff
@@ -23,3 +23,152 @@ distribution:
     notification_emails:
       - dmu@mozilla.com
     expires: "2020-05-01"
+pages:
+  page_load:
+    type: timing_distribution
+    lifetime: application
```
Comment (Dexterp37) [inline app/metrics.yaml]: Are you sure this is really an `application` lifetime metric? This sounds like a `ping` lifetime metric: you really want the metric to be cleared after the ping is sent, not when the application is shut down.

File: app/metrics.yaml [hunk 4]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+url:
+  domains:
+    type: counter
+    lifetime: application
+    description: >
+      Counting how many domains are visited in a session.
+    send_in_pings:
+      - metrics
```
Comment (Dexterp37) [inline app/metrics.yaml]: Redundant :)

File: app/metrics.yaml [hunk 5]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
+    notification_emails:
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+  visits:
+    type: counter
+    lifetime: application
+    description: >
+      Counting how many URL links are visited in a session.
+    send_in_pings:
```
Comment (Dexterp37) [inline app/metrics.yaml]: Redundant :)

File: app/metrics.yaml [hunk 6]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+  insert_types:
+    type: labeled_counter
+    lifetime: application
+    description: >
+      Counting how many domains and URL are visited in a session.
+    labels:
+      - type_link
+      - type_query
+      - voice_query
+    send_in_pings:
```
Comment (Dexterp37) [inline app/metrics.yaml]: Redundant :)

File: app/metrics.yaml [hunk 7]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+    notification_emails:
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+durarion:
+  immersive_mode:
+    type: timing_distribution
+    lifetime: application
+    description: >
+      Counting how long users stay at the immersive mode.
+    send_in_pings:
```
Comment (Dexterp37) [inline app/metrics.yaml]: Redundant :)

File: app/metrics.yaml [hunk 8]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
+    notification_emails:
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+  window_lifetime:
+    type: timing_distribution
+    lifetime: application
+    description: >
+      Counting how long is a Window opened.
+    send_in_pings:
```
Comment (Dexterp37) [inline app/metrics.yaml]: Redundant :)

File: app/metrics.yaml [hunk 9]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+    notification_emails:
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+controls:
+  tray_new_window:
+    type: counter
+    lifetime: application
+    description: >
+      Counting how many times users open new windows.
+    send_in_pings:
```
Comment (Dexterp37) [inline app/metrics.yaml]: Redundant :)

File: app/metrics.yaml [hunk 10]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
+    notification_emails:
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+  context_menu_long_pressed:
+    type: counter
+    lifetime: application
+    description: >
+      Counting how many times users do a long press on the context menu.
+    send_in_pings:
```
Comment (Dexterp37) [inline app/metrics.yaml]: Redundant :)

File: app/metrics.yaml [hunk 11]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
+    notification_emails:
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+  curve_mode_actived:
+    type: boolean
+    lifetime: application
+    description: >
+      Counting how many users using the curved mode.
+    send_in_pings:
```
Comment (Dexterp37) [inline app/metrics.yaml]: Redundant :)

File: app/metrics.yaml [hunk 12]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+    notification_emails:
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+searches:
+  engine:
+    type: labeled_counter
+    lifetime: application
+    description: >
+      Counting how many searches are queried in a specific search engine.
+    send_in_pings:
```
Comment (Dexterp37) [inline app/metrics.yaml]: Redundant :)

File: app/metrics.yaml [hunk 13]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+      - metrics
+    bugs:
+      - https://github.com/MozillaReality/FirefoxReality/issues/2230
+    data_reviews:
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
+    notification_emails:
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+searches:
+  engine:
```
Comment (Dexterp37) [inline app/metrics.yaml]: nit: this is not really an engine, this is the number of searches. Maybe `counts` is a better name?

File: app/metrics.yaml [hunk 14]
```diff
@@ -23,3 +23,152 @@ distribution:
     notification_emails:
       - dmu@mozilla.com
     expires: "2020-05-01"
+pages:
+  page_load:
```
Comment (Dexterp37) [inline app/metrics.yaml]: nit: `load_time`?

File: app/metrics.yaml (hunk 3 above)
Comment (Dexterp37) [reply app/metrics.yaml]: Note that the lifetime consideration also applies to all the other metrics below.

File: app/metrics.yaml [hunk 15]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+    type: timing_distribution
+    lifetime: application
+    description: >
+      Counting how long each page takes to load.
```
Comment (Dexterp37) [inline app/metrics.yaml]: Please be a bit more specific about what you are measuring: does this include DNS lookup? What does "page load" mean? How is a page considered fully loaded? If you are unsure, you can mention that this relies on the `onPageStart` and `onPageStop` from GeckoView. You can also link to them using Markdown: this will be linkified when generating the docs.

File: app/metrics.yaml [hunk 16]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
+    notification_emails:
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+url:
+  domains:
+    type: counter
+    lifetime: application
+    description: >
+      Counting how many domains are visited in a session.
```
Comment (Dexterp37) [inline app/metrics.yaml]: What is a session? Are the domains unique?

File: app/metrics.yaml [hunk 17]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+    data_reviews:
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
+    notification_emails:
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+  visits:
+    type: counter
+    lifetime: application
+    description: >
+      Counting how many URL links are visited in a session.
```
Comment (Dexterp37) [inline app/metrics.yaml]: What is a session? :)

File: app/src/common/shared/org/mozilla/vrbrowser/VRBrowserActivity.java [hunk 18]
```diff
@@ -380,9 +381,13 @@ protected void onStop() {
 
         if (SettingsStore.getInstance(this).getCylinderDensity() > 0.0f) {
             TelemetryWrapper.queueCurvedModeActiveEvent();
+            GleanMetricsService.curveModeActiveEvent(true);
```
Comment (Dexterp37) [inline app/src/common/shared/org/mozilla/vrbrowser/VRBrowserActivity.java]: Does this buy you anything over just calling the Glean APIs? (and in other places as well)

File: app/metrics.yaml [hunk 19]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+    send_in_pings:
+      - metrics
+    bugs:
+      - https://github.com/MozillaReality/FirefoxReality/issues/2230
+    data_reviews:
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
+    notification_emails:
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+  insert_types:
```
Comment (Dexterp37) [inline app/metrics.yaml]: is `query_type` or `query_input` a better name?

File: app/metrics.yaml [hunk 20]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+    type: labeled_counter
+    lifetime: application
+    description: >
+      Counting how many domains and URL are visited in a session.
```
Comment (Dexterp37) [inline app/metrics.yaml]:
```suggestion
      Counting how many domains and URL are visited in a session, by input method.
```

File: app/metrics.yaml [hunk 21]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+    send_in_pings:
+      - metrics
+    bugs:
+      - https://github.com/MozillaReality/FirefoxReality/issues/2230
+    data_reviews:
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
+    notification_emails:
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+durarion:
```
Comment (Dexterp37) [inline app/metrics.yaml]: I think this should really be the other way around: `immersive.mode: duration`. We [discourage](https://mozilla.github.io/glean/book/user/adding-new-metrics.html#naming-things) using "types" as category names.

File: app/metrics.yaml [hunk 22]
```diff
@@ -23,3 +23,152 @@ distribution:
...
+    send_in_pings:
+      - metrics
+    bugs:
+      - https://github.com/MozillaReality/FirefoxReality/issues/2230
+    data_reviews:
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
+    notification_emails:
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+  window_lifetime:
```
Comment (Dexterp37) [inline app/metrics.yaml]: What about a `window` category with a `opened_duration` or something like that?

**Comment by Dexterp37** [comment]
> Blocked by [Bug 1595914](https://bugzilla.mozilla.org/show_bug.cgi?id=1595914) to access `GleanTimerId`

@daoshengmu I think the rest of the metrics can be landed before  [Bug 1595914](https://bugzilla.mozilla.org/show_bug.cgi?id=1595914) is fixed. Just remove the ones that need a timer id from this PR and create a new PR once that bug is fixed :)

Before landing new metrics, did you verify (as [stated by the migration docs](https://docs.google.com/document/d/1k7GabrWCDKp7Ae8mqr-XFnFxIuNlxsJM52D-1uVmgb0/edit?ts=5d8aadbd)) that the `baseline` ping is flowing in and both its volume and its data is consistent to what you're receiving with `service-telemetry`? Do you need help with that?

Given the `WorkManager`  usage, did you check if things are flowing in from the platform that had permission problems?

File: app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java (hunk 1 above)
Comment (daoshengmu) [reply app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java]: One session means the duration of this application is running at the foreground. So, while the application switch to the background, our application will receive an `onStop()` event, and that means the current session has been ended. The session lifetime sounds like very close to `lifetime: application`? if it is not totally the same, I guess I still can use `lifetime: ping` to replace it.

**Comment by daoshengmu** [comment]
> > Blocked by [Bug 1595914](https://bugzilla.mozilla.org/show_bug.cgi?id=1595914) to access `GleanTimerId`
> 
> @daoshengmu I think the rest of the metrics can be landed before [Bug 1595914](https://bugzilla.mozilla.org/show_bug.cgi?id=1595914) is fixed. Just remove the ones that need a timer id from this PR and create a new PR once that bug is fixed :)
> 
> Before landing new metrics, did you verify (as [stated by the migration docs](https://docs.google.com/document/d/1k7GabrWCDKp7Ae8mqr-XFnFxIuNlxsJM52D-1uVmgb0/edit?ts=5d8aadbd)) that the `baseline` ping is flowing in and both its volume and its data is consistent to what you're receiving with `service-telemetry`? Do you need help with that?
> 
> Given the `WorkManager` usage, did you check if things are flowing in from the platform that had permission problems?

Our FxR version 6 hasn't officially released yet because other blocker issues. We expect one day next week. You can check the correct status at here, https://sql.telemetry.mozilla.org/dashboard/firefox-reality-telemetrycmp.

File: app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java (hunk 1 above)
Comment (daoshengmu) [reply app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java]: I think I can use custom pings, `session_end`, to help me send these metrics once application jumps to the background.

File: app/metrics.yaml (hunk 15 above)
Comment (daoshengmu) [reply app/metrics.yaml]: I am going to move this to anther PR because of blocker issues.

File: app/src/common/shared/org/mozilla/vrbrowser/VRBrowserActivity.java (hunk 18 above)
Comment (daoshengmu) [reply app/src/common/shared/org/mozilla/vrbrowser/VRBrowserActivity.java]: It's just for easy maintenance.

File: app/metrics.yaml (hunk 22 above)
Comment (daoshengmu) [reply app/metrics.yaml]: I am going to move this to another PR because of blockers.

File: app/metrics.yaml (hunk 21 above)
Comment (daoshengmu) [reply app/metrics.yaml]: Let's fix it at anther PR.

**Comment by daoshengmu** [comment]
> Blocked by [Bug 1595914](https://bugzilla.mozilla.org/show_bug.cgi?id=1595914) to access `GleanTimerId`

I have move the blocked metrics to #2279.

File: app/pings.yaml [hunk 23]
```diff
@@ -0,0 +1,17 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+$schema: moz://mozilla.org/schemas/glean/pings/1-0-0
+
+session_end:
+  description: >
+    These pings are sent at the end of a session. (when Firefox Reality switches to the background.)
```
Comment (Dexterp37) [inline app/pings.yaml]:
```suggestion
    This ping is sent at the end of a session (when Firefox Reality switches to the background).
```

File: app/pings.yaml [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+    We usually send search and UI control metrics at the end of a session.
+  include_client_id: false
```
Comment (Dexterp37) [inline app/pings.yaml]: Are you sure you don't want the `client_id` here? I think it would make your ping useless if you don't have it :)

File: app/pings.yaml [hunk 25]
```diff
@@ -0,0 +1,17 @@
...
+  bugs:
+    - https://github.com/MozillaReality/FirefoxReality/issues/2230
+  data_reviews:
+    - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
```
Comment (Dexterp37) [inline app/pings.yaml]: Please request a new data review for this ping and all the metrics in this PR.

File: app/pings.yaml [hunk 26]
```diff
@@ -0,0 +1,17 @@
...
+  notification_emails:
+    - dmu@mozilla.com
```
Comment (Dexterp37) [inline app/pings.yaml]: Does Firefox REality have a team mailing list or so? In this case, you should probably add it both here and to the metrics.

File: app/metrics.yaml [hunk 27]
```diff
@@ -23,3 +23,100 @@ distribution:
     notification_emails:
       - dmu@mozilla.com
     expires: "2020-05-01"
+url:
+  domains:
+    type: counter
+    lifetime: ping
```
Comment (Dexterp37) [inline app/metrics.yaml]: Note: `lifetime: ping` is redundant and can be removed here and elsewhere below. It's the default lifetime.

File: app/metrics.yaml [hunk 28]
```diff
@@ -23,3 +23,100 @@ distribution:
...
+    send_in_pings:
+      - session_end
+    description: >
+      Counting how many domains are visited in a session.
+    bugs:
+      - https://github.com/MozillaReality/FirefoxReality/issues/2230
+    data_reviews:
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
```
Comment (Dexterp37) [inline app/metrics.yaml]: This and all the new metrics in this file require data review.

File: app/metrics.yaml [hunk 29]
```diff
@@ -23,3 +23,100 @@ distribution:
...
+      - type_query
+      - voice_query
+    bugs:
+      - https://github.com/MozillaReality/FirefoxReality/issues/2230
+    data_reviews:
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
+    notification_emails:
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+controls:
```
Comment (Dexterp37) [inline app/metrics.yaml]: From this point on all metrics are missing the `send_in_pings` property and won't be sent in your newly defined ping.

File: docs/metrics.md [hunk 30]
```diff
@@ -38,7 +38,17 @@ The following metrics are added to the ping:
 
 | Name | Type | Description | Data reviews | Extras | Expiration |
 | --- | --- | --- | --- | --- | --- |
+| controls.context_menu_long_pressed |[counter](https://mozilla.github.io/glean/book/user/metrics/counter.html) |Counting how many times users do a long press on the context menu.  |[1](https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568)||2020-05-01 |
```
Comment (Dexterp37) [inline docs/metrics.md]: This is out of date. Did you commit after you added `pings.yaml`? It should have documented the changes here.

File: app/pings.yaml [hunk 31]
```diff
@@ -0,0 +1,17 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+$schema: moz://mozilla.org/schemas/glean/pings/1-0-0
+
+session_end:
```
Comment (Dexterp37) [inline app/pings.yaml]: Since you added a new ping, you should also add your `distribution_name` metric to this one.

**Comment by Dexterp37** [comment]
@daoshengmu have you considered adding unit tests for your key metrics to make sure they don't regress?

File: app/pings.yaml (hunk 24 above)
Comment (daoshengmu) [reply app/pings.yaml]: The `client_id` is the one I am not very sure. I was thinking I might not need it if I only wanna accumulate counters daily instead of recognizing these pings by `client_id` for per user. However,  including client id absolutely gives us more flexible.

File: app/pings.yaml (hunk 26 above)
Comment (daoshengmu) [reply app/pings.yaml]: Interesting! I was adding our team mailing list when doing the first stage migrating. IIRC, you would like me to use my personal mail? If you prefer to use to the team mail list, I think I can just put the team's one here because that group already include me.

File: app/metrics.yaml (hunk 29 above)
Comment (daoshengmu) [reply app/metrics.yaml]: Those are discussing with my team. They haven't decided `tray_new_window` and `context_menu_long_pressed` counters if they should send at the end of a session and can get a reset in the new session. Or,  they wanna count them daily, so they would be ping lifetime.

`curve_mode_actived` is a boolean, so it can  be sent at its ping lifetime for sure.

File: app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java [hunk 32]
```diff
@@ -53,6 +63,103 @@ public static void stop() {
...
+
+    @UiThread
+    public static void urlBarEvent(boolean aIsUrl) {
+        if (aIsUrl) {
+            Url.INSTANCE.getQueryType().get("type_link").add();
+        } else {
+            Url.INSTANCE.getQueryType().get("type_query").add();
+            // Record search engines.
+            String searchEngine = getDefaultSearchEngineIdentifierForTelemetry();
+            Searches.INSTANCE.getCounts().get(searchEngine).add();
```
Comment (daoshengmu) [inline app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java]: This seems to not work in Glean Debug Ping Viewer. `searchEngine` is a search engine like `www.google.com`, but in Viewer, it shows `invalid_label`.

**Comment by daoshengmu** [comment]
One thing I might miss is 

> @daoshengmu have you considered adding unit tests for your key metrics to make sure they don't regress?

Yes. I will do some study on Monday although FxR seems not have a fundamental work  for unit testing, and I will request the data review soon.

File: app/pings.yaml (hunk 24 above)
Comment (Dexterp37) [reply app/pings.yaml]: I'm fairly confident you need the `client_id` here, as you had that in your previous pings.

File: app/metrics.yaml (hunk 29 above)
Comment (Dexterp37) [reply app/metrics.yaml]: > 
> 
> Those are discussing with my team. They haven't decided `tray_new_window` and `context_menu_long_pressed` counters if they should send at the end of a session and can get a reset in the new session. Or, they wanna count them daily, so they would be ping lifetime.
> 
> `curve_mode_actived` is a boolean, so it can be sent at its ping lifetime for sure.

Please note that the fact that a metric is `lifetime: ping` has nothing to do with session: if you add the metric to your custom ping being send in `onStop`, then metrics will be cleared right after. If you add them to the `metrics` ping, then they will be cleared when that ping is sent, once per day. So it really depends on the ping.

I would recommend to keep things consistent, otherwise you'd need to two different analyses types.

File: app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java (hunk 32 above)
Comment (Dexterp37) [reply app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java]: According to your [implementation of `getDefaultSearchEngineIdentifierForTelemetry`](https://github.com/MozillaReality/FirefoxReality/pull/2241/files#diff-cb02bef369d942cc3f925e33897d6fe7R159), you're using `getResourceURL` which returns a [full URL](https://github.com/MozillaReality/FirefoxReality/blob/7fcabcb2834bbbdf51cafb9a17a58271f6c95cbd/app/src/common/shared/org/mozilla/vrbrowser/search/SearchEngineWrapper.java#L125) for a search engine. As you can see from our docs, [labels only accept specific set of characters](https://mozilla.github.io/glean/book/user/metrics/labeled_counters.html#limits).

[Fenix](https://github.com/mozilla-mobile/fenix/blob/c6562bff98d602b969a71fba3dc84b86821251d4/app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt#L514) uses the [lower case `identifier`](https://github.com/mozilla-mobile/fenix/blob/c6562bff98d602b969a71fba3dc84b86821251d4/app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt#L269) as a label for counts.

You should probably do the same.

**Comment by Dexterp37** [comment]
> 
> 
> One thing I might miss is
> 
> > @daoshengmu have you considered adding unit tests for your key metrics to make sure they don't regress?
> 
> Yes. I will do some study on Monday although FxR seems not have a fundamental work for unit testing, and I will request the data review soon.

Great! Let us know if you need any help with writing your first unit test for telemetry.

File: app/metrics.yaml (hunk 29 above)
Comment (daoshengmu) [reply app/metrics.yaml]: After the discussion with our team, we are going to remove these UI control telemetry in the following version, so we don't need to support them.

**Comment by daoshengmu** [comment]
# Request for data collection review form
 **All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**
 
 1. **What questions will you answer with this data?**
 
 *  How many domains are visited in a session?
 *  How many URLs are visited in a session?
 *  How many times do users insert URL, search query, and voice query in the URL bar in a session?
 *  How many searches are queried in a specific search engine in a session?
 
 2. **Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?**
 
 * To measure how users insert URL into the URL bar, by inserting? query? or voice input? We could provide greater experience for searches.
 * We wanna understand which search engine is the most users would like to use.
 * We also are interested in analyzing users' behavior during a session about how often they use search engine and the search bar.
  
 3. **What alternative methods did you consider to answer these questions? Why were they not sufficient?**
 
 * N/A
 
 4. **Can current instrumentation answer these questions?**
 
 * Currently no, as these are some of the first metrics we're recording.
 
 5. **List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox [data collection](https://wiki.mozilla.org/Firefox/Data_Collection) [categories](https://wiki.mozilla.org/Firefox/Data_Collection#Data_Collection_Categories) on the found on the Mozilla wiki.**
 
 * All data is Category 2.
 
 6. **How long will this data be collected?**
 
 Until 05/01/2020
 
 7. **What populations will you measure?**

 * All release, beta, and nightly users with telemetry enabled.
 
 8. **Please provide a general description of how you will analyze this data.**
 
 * Glean
 
 9. **Where do you intend to share the results of your analysis?**
 
 * Only on Glean and with Firefox Reality team..

**Comment by daoshengmu** [comment]
@harraton Please help [data review](https://github.com/MozillaReality/FirefoxReality/pull/2241#issuecomment-555285627).

File: app/metrics.yaml [hunk 33]
```diff
@@ -16,10 +16,72 @@ distribution:
...
+    bugs:
+      - https://github.com/MozillaReality/FirefoxReality/issues/2230
+    data_reviews:
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
+    notification_emails:
+      - fxr-telemetry@mozilla.com
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+searches:
+  counts:
```
Comment (Dexterp37) [inline app/metrics.yaml]: This is lacking the `send_in_pings` and it's not being sent in your new ping.

File: app/src/common/shared/org/mozilla/vrbrowser/search/SearchEngineWrapper.java [hunk 34]
```diff
@@ -125,6 +125,11 @@ public String getResourceURL() {
         return uri.getScheme() + "://" + uri.getHost();
     }
 
+    public String getHostURL() {
+        Uri uri = Uri.parse(mSearchEngine.buildSearchUrl("")) ;
+        return uri.getHost();
```
Comment (Dexterp37) [inline app/src/common/shared/org/mozilla/vrbrowser/search/SearchEngineWrapper.java]: I understand this is _your_ instrumentation and your data but please consider that all the other products (including Firefox Desktop, Fennec and Fenix) are using the engine `identifier` rather than the host. This doesn't block the review, but you might want to consider using the same.

File: app/metrics.yaml (hunk 16 above)
Comment (daoshengmu) [reply app/metrics.yaml]: We use a hashSet to identify different domains, but we are not interested in sending domain names to the data server. We only send hashSet().size().

File: app/metrics.yaml (hunk 33 above)
Comment (daoshengmu) [reply app/metrics.yaml]: I was thinking the searches.count can be sent in metrics. We are interested in collecting it in daily instead of by sessions, and it will go by default into the metrics ping. Therefore, I ignored `send_in_pings`, does it make sense?

Comment (Dexterp37) [reply app/metrics.yaml]: It does, I wanted to make sure that was your intent ;) Since that's the intent, it's fine!

File: app/src/common/shared/org/mozilla/vrbrowser/search/SearchEngineWrapper.java (hunk 34 above)
Comment (daoshengmu) [reply app/src/common/shared/org/mozilla/vrbrowser/search/SearchEngineWrapper.java]: I got `google-b-1-m` from the `SearchEngine.identifier`. Is this what we want? I can follow what they did, no problem.

File: app/metrics.yaml [hunk 35]
```diff
@@ -16,10 +16,70 @@ distribution:
...
+      Counting how many URL links are visited in a session.
+    bugs:
+      - https://github.com/MozillaReality/FirefoxReality/issues/2230
+    data_reviews:
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
+    notification_emails:
+      - fxr-telemetry@mozilla.com
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+  query_type:
```
Comment (daoshengmu) [inline app/metrics.yaml]: Our team would like to make `query_type` to be `lifetime:ping` and send it daily. @Dexterp37 Do you think if I need to move it out of `url` category because others are sent in `session_end`?

Comment (Dexterp37) [reply app/metrics.yaml]: No, I think it's fine to leave it in that category, no worries! Categories are meant as a way to organize semantically related data, so no worries! Things with different lifetimes can be under the same category.

That's really up to you and how do you want this file, _your file_, to be structured :) One question that might inform your choice is the following: "Would I change the name of the category or the metric if I changed the lifetime?". In general, you would want the metric identifier (category + name) to be about what the metric is, not when the metric gets cleared.

So, to summarise, my personal opinion would be to leave the metric in this category, just changing the lifetime.

File: app/src/common/shared/org/mozilla/vrbrowser/search/SearchEngineWrapper.java (hunk 34 above)
Comment (Dexterp37) [reply app/src/common/shared/org/mozilla/vrbrowser/search/SearchEngineWrapper.java]: Yes! I think that is exactly what you want. For more context see [the issue](https://github.com/mozilla-mobile/fenix/issues/4082#issuecomment-512187680) that Fenix fixed a bit back.

I'd recommend sticking to what Fenix is doing since they will be improving that part a bit to support mozilla-mobile/fenix#6448

File: app/metrics.yaml [hunk 36]
```diff
@@ -16,10 +16,70 @@ distribution:
...
+      - https://github.com/MozillaReality/FirefoxReality/pull/1854#issuecomment-546214568
+    notification_emails:
+      - fxr-telemetry@mozilla.com
+      - dmu@mozilla.com
+    expires: "2020-05-01"
+searches:
+  counts:
+    type: labeled_counter
+    description: >
+      Counting how many searches are queried in a specific search engine.
```
Comment (Dexterp37) [inline app/metrics.yaml]:
```suggestion
      Counting how many searches are queried in a specific search engine. The search engine `identifier`s are used as keys for this metric.
```

**Comment by keianhzo** [comment]
@daoshengmu Ready to merge when @harraton completes the data review.

File: app/metrics.yaml (hunk 28 above)
Comment (Dexterp37) [reply app/metrics.yaml]: Let's not forget to update the `data_reviews` fields before merging :)

**Comment by harraton** [comment]
Data review approved. 

# Data Review Form

Instructions: Data Stewards will review a request for data collection and endorse responses to each question. **If the request does not provide answers to questions, reviewers give an r- and point to the questions that can't be answered.**

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way? 
Yes: 
https://github.com/MozillaReality/FirefoxReality/wiki/Telemetry#new-telemetry-system---glean

2) Is there a control mechanism that allows the user to turn the data collection on and off? 
Yes, on the settings page for Firefox Reality. 

3) If the request is for permanent data collection, is there someone who will monitor the data over time?
Yes, @daoshengmu will monitor. 

4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?
Category 2

5) Is the data collection request for default-on or default-off?
Default on

6) Does the instrumentation include the addition of **any *new* identifiers** 
No.

7) Is the data collection covered by the existing Firefox privacy notice?
Covered by the Firefox Reality Privacy Policy

8) Does there need to be a check-in in the future to determine whether to renew the data? (Yes) In May 2020. 

9) Does the data collection use a third-party collection tool? 
Yes, Glean.
